lld/ELF/ICF.cpp

```diff
@@ -247,11 +250,21 @@ bool ICF<ELFT>::constantEq(const InputSection *secA, Relocs<RelTy> ra,
         rai->getType(ctx.arg.isMips64EL) != rbi->getType(ctx.arg.isMips64EL))
       return false;
 
-    uint64_t addA = getAddend<ELFT>(*rai);
-    uint64_t addB = getAddend<ELFT>(*rbi);
+    int64_t addA = getAddend<ELFT>(*rai);
+    int64_t addB = getAddend<ELFT>(*rbi);
 
     Symbol &sa = secA->file->getRelocTargetSym(*rai);
     Symbol &sb = secB->file->getRelocTargetSym(*rbi);
+
+    // We only need relocation type and relocation target symbol. Offset is
+    // not used (as of now). So we don't bother populating one.
+    relocsA.push_back(
+        {ctx.target->getRelExpr(rai->getType(ctx.arg.isMips64EL), sa, 0),
+         rai->getType(ctx.arg.isMips64EL), 0, addA, &sa});
+    relocsB.push_back(
+        {ctx.target->getRelExpr(rbi->getType(ctx.arg.isMips64EL), sb, 0),
+         rbi->getType(ctx.arg.isMips64EL), 0, addB, &sb});
```

Comment:
happy to either only pass rel.type and Symbol to TargetInfo::canFoldSection but passing `Relocation` sounded better to me.